src/server/proxy_protocol.rs

```diff
@@ -0,0 +1,203 @@
+use super::session::SharedSession;
+use crate::auth::UserDetail;
+use crate::storage;
+
+use bytes::Bytes;
+use lazy_static::*;
+use log::warn;
+use proxy_protocol::version1::ProxyAddressFamily;
+use proxy_protocol::ProxyHeader;
+use rand::rngs::OsRng;
+use rand::RngCore;
+use std::collections::HashMap;
+use std::net::IpAddr;
+use std::ops::Range;
+use tokio::io::AsyncReadExt;
+use tokio::sync::Mutex;
+
+lazy_static! {
+    static ref OS_RNG: Mutex<OsRng> = Mutex::new(OsRng);
+}
+
+#[derive(Debug)]
+pub enum ProxyError {
+    CrlfError,
+    HeaderSize,
+    NotProxyHdr,
+    DecodeError,
+    IPv4Required,
+    UnsupportedVersion,
+}
+
+#[derive(Debug, Copy, Clone)]
+pub struct ConnectionTuple {
+    pub from_ip: IpAddr,
+    pub from_port: u16,
+    pub to_ip: IpAddr,
+    pub to_port: u16,
+}
+
+impl ConnectionTuple {
+    fn new(from_ip: IpAddr, from_port: u16, to_ip: IpAddr, to_port: u16) -> Self {
+        ConnectionTuple {
+            from_ip,
+            from_port,
+            to_ip,
+            to_port,
+        }
+    }
+}
+
+async fn read_proxy_header(tcp_stream: &mut tokio::net::TcpStream) -> Result<ProxyHeader, ProxyError> {
```

Comment:
We should probably write some unit tests for this after we've merged this.

Reply:
yes..... that was in my mind as well..... :)